homeassistant/helpers/event.py

```diff
@@ -90,34 +99,246 @@ def state_change_listener(event):
+        entity_id = event.data.get('entity_id')
+        # Optimisation: if the old and new states are not None then this is
+        # a state change rather than a life-cycle event, and therefore we
+        # only need to check the include entities.
+        old_state = event.data.get('old_state')
+        new_state = event.data.get('new_state')
+        if old_state is not None and new_state is not None:
+            if not self._info.filter(entity_id):
+                return
+        elif not self._info.filter_lifecycle(entity_id):
+            return
+
+        self._refresh(event)
 
-        # Check to see if template returns true
-        if template_result and not already_triggered:
-            already_triggered = True
-            hass.async_run_job(action, entity_id, from_s, to_s)
-        elif not template_result:
-            already_triggered = False
+    def _refresh(self, event=None) -> None:
+        self._info = self._template.async_render_to_info(self._variables)
 
-    return async_track_state_change(
-        hass, template.extract_entities(variables),
-        template_condition_listener)
+        try:
+            result = self._info.result
+        except TemplateError as ex:
+            if not self._last_exception:
+                self.hass.async_run_job(
+                    self._action, event, self._template,
+                    self._last_result, ex)
+            self._last_exception = True
+            return
+        self._last_exception = False
 
+        # Check to see if the result has changed
+        if result == self._last_result:
+            return
 
-track_template = threaded_listener_factory(async_track_template)
+        self.hass.async_run_job(
+            self._action, event, self._template,
+            self._last_result, result)
+        self._last_result = result
+
+
+TrackTemplateResultListener = Callable[[
+    Optional[Event],
+    Template,
+    Optional[str],
+    Union[str, TemplateError]], None]
+"""Type for the listener for template results.
+
+    Action arguments
+    ----------------
+    event
+        Event that caused the template to change output. None if not
+        triggered by an event.
+    template
+        The template that has changed.
+    last_result
+        The output from the template on the last successful run, or None
+        if no previous successful run.
+    result
+        Result from the template run. This will be a string or an
+        TemplateError if the template resulted in an error.
+"""
+
+
+@callback
+@bind_hass
+def async_track_template_result(
+        hass: HomeAssistant,
+        template: Template,
+        action: TrackTemplateResultListener,
+        variables: Optional[Mapping[str, Any]] = None) \
+        -> TrackTemplateResultInfo:
+    """Add a listener that fires when a the result of a template changes.
+
+    The action will fire with the initial result from the template, and
```

Comment:
Why are we firing it with the initial result?

Reply:
See above comment